Articles/Blog/WhenIsBetterToWriteAutomatedTests-src.md

```diff
@@ -0,0 +1,110 @@
+
+*SOME GOOD SUBTITLE.  Any suggestions?*
+
+<br>
+
+## Introduction
+
+The importance of writing and maintaining strong automated tests for software is well established in the modern software engineering community (and in fact, tests will be written before the code when using test-driven development, TDD).
+But does it always pay off to write automated tests for some piece of code?
+Are there situations where one is better off not writing automated tests?
+Will you be considered to be a bad developer and chastised by your peers, users, stakeholders, or your manager if every line of code that you write is not under strong automated testing?
+
+I have been wrestling with this question since 2007 when I first read the book "Working Effectively with Legacy Code"<sup>[1]</sup> and radically changed my thinking about automated testing.
+In the 10+ years of my professional career prior to 2007, I wrote a lot of Computational Science and Engineering (CSE) and applied math software with very little strong automated testing.
+I used the validation-centric approach<sup>[2]</sup> to test the software that I was writing (i.e. only test the software in the downstream customer code and then parhaps only manually).
+As a result, I had personally written O(100k) lines of code that were being used by other people and was in the position to have to maintain.
+When I realized that almost all of this software was considered "Legacy Code" according to "Working Effectively with Legacy Code" (i.e. "code without tests == legacy code"), I felt ashamed and horrified at the mass of useful but buggy and difficult-to=maintain software that I had created and now would have to deal with for years to come.
+(Just recently a nasty bug came up in software I wrote around 2006 without strong automated testing that caused a good deal of harm and was an embarrassment.)
+
+After reading the book "Working Effectively with Legacy Code", I resolved to turn over a new leaf and become a born-again Agile software engineer that would use TDD to write all new code and religiously apply the Legacy Code Change Algorithm to modify all legacy code (i.e. code without tests).
+Over the years since my conversion to an Agile software engineer, I often rigorously wrote automated tests for nearly every piece of software I touched.
+There where times  under schedule pressure that I slipped and reverted back to my old ways and failed to write sufficient automated tests and came to regret it.
+However, there were also times where my overzealous drive to religiously test everything (which usually came after the guilt of a recent slip of not testing enough) where I wasted time writing automated tests that were not worth the effort.
+And now after having had followed this road from 2007 until now having written and maintained thousands of automated tests over that time, I have come to realize that there are times when it is actually better not to write automated tests for some pieces of software!
```

Comment:
And now after having had followed this road from 2007 until now having written and maintained thousands of automated tests over that time, I have come to realize that while majority of the times automated tests are necessary and critical, there are times when it is actually better not to write automated tests for some pieces of software!